lib/std/fmt/scan.zig

```diff
@@ -0,0 +1,239 @@
+const std = @import("std");
+
+pub const Error = error{ InputMatchFailure, TooFewArguments, TooManyArguments, UnknownSpecifier, NegationOfSignedInteger, ConversionFailure, InvalidCharacter };
+
+fn parseInt(input: []const u8, radix: u8, isValid: fn (u8) bool, value: anytype, neg: bool) Error!usize {
+    std.debug.assert(@typeInfo(@TypeOf(value)) == .Pointer);
+    const Int = @TypeOf(value.*);
+    if (neg and !@typeInfo(Int).Int.is_signed) return error.NegationOfSignedInteger;
+
+    var i: usize = 0;
+    const radix_adj = @intCast(Int, radix);
+
+    value.* = 0;
+    while (i < input.len) : (i += 1) {
+        const c = input[i];
+        if (!isValid(c)) break;
+        value.* = value.* * radix_adj + @intCast(Int, try std.fmt.charToDigit(c, radix));
+    }
+    if (@typeInfo(Int).Int.is_signed) {
+        if (neg) value.* = -value.*;
+    }
+    if (i == 0) return error.ConversionFailure;
+    return i;
+}
+
+fn isOctalDigit(c: u8) bool {
+    return '0' <= c and c <= '7';
+}
+
+fn isBinaryDigit(c: u8) bool {
+    return c == '0' or c == '1';
+}
+
+/// similar to scanf in c except conversion specifiers are given like "{d}" rather than "%d"
+/// valid specifiers include
+/// - {c}     - match exactly one character, don't parse. '0' -> '0' (48)
+/// - {d}, {} - match and parse a base 10 number
+/// - {e}     - match and parse a floating point number
+/// - {x}     - match and parse a hex number with optional leading "0x" or "0X"
+/// - {o}     - match and parse an octal number with optional leading "0o"
+/// - {b}     - match and parse a binary number with optional leading "0b"
+/// - {s}     - match any input until a whitespace or the following character in fmt string
+/// Notes: curly braces can be escaped in fmt "{{}}"
+///        each whitespace in fmt will skip zero or more input whitespaces
+pub fn scan(input: []const u8, comptime fmt: []const u8, args: anytype) Error!usize {
+    comptime var arg_i: usize = 0;
+    var i: usize = 0; // input index
+    const State = enum { start, specifier, specifier_end };
+    comptime var state: State = .start;
+    inline for (fmt) |fmt_c, fmt_i| {
+        var neg = false;
+        if (i == std.math.maxInt(usize)) return error.InputMatchFailure;
+        switch (state) {
```

Comment:
Thank you for the reviews.  This seems like the best way forward to me, extracting and re-using the existing parser.  I'll take a closer look at that pr.  Let me know if you have an idea of what this might look like once its extracted.  A struct with method fields?  Or maybe just fmt.format() but with method argument?  I hope to work more on this tomorrow.

Reply:
Well, it depends on what the specification for `scan` is. I doubt the `alignment` and `fill` specifiers are needed and the only place where I can see a width specifier to be useful is with the `c` specifier.
If you can live with this reduced set of options I'd simply extract the `parser` (and `arg_state`) struct and all you have to do is copy-paste the `inline while` in `format` and modify it to suit your needs.

Reply:
I just pushed a commit that addresses many of these points.  Haven't handled precision yet, but initial width support is there. There are many TODOs in the code.  For one, I wasn't sure what to do about if {c} has a width or precision.